### ISSUE TITLE
Potential Bugs in loss function and TransformerDecoder layers

### ISSUE THREAD

**Issue by noorahmedds** [elisabettafedele/superdec#2]
Hi authors,

Firstly, thanks for sharing the code for the great work. Congratulations on the ICCV acceptance. I was playing around in the repository, and I saw you could fix some hard-to-see bugs that might improve your results further. 

1. The cd_loss returns only the one-way "**pcl_to_prim_loss**" and the other way is unused in code as described in the paper. 
2. A single decoder layer is simply repeated during construction in the TransformerDecoder, which means the decoder layers essentially **share the same weight** (this seems non-standard to me). Is this intended? If not, you may want to instantiate a list of layers in superdec.py and then pass it into the TransformerDecoder constructor as a starting point

I'll drop a _pull request_ for my suggestions as reference, and you can see if these help you out. I have a strong feeling they will :)

**Comment by elisabettafedele** [comment]
Hi Noor,

Thank you so much for your comments and for your interest in our work! I will definitely update the loss (unfortunately this is an error introduced during refactoring) and try to train a new version with different transformer layers. I still want to wait to incorporate suggestion #2 because it would not be consistent with the current shared checkpoints :)

Have the modifications improved the results in your case?

Best,
Elisabetta

**Comment by noorahmedds** [comment]
Hi Elisabetta,

Indeed the results did improve; however, the following are runs without LM-Optimization. I also simply used the default lambda_parsimony weight defined in the yaml, which is different from the paper (likely the reason for the high compactness), but here are the two runs without and with the fixes, respectively:

CD L1| CD L2 | Compactness 
1.7130 | 0.0527 | 10.635`
1.6451 | 0.0432 | 10.242

Unsure if the impact would be as significant when you set the parsimony loss weight to 0.6. You might need to search for the optimal loss weighting again.  

FYI: Making deep copies of the decoder layers seems more standard than the solution I suggested. [Here](https://github.com/elisabettafedele/superdec/blob/806afd2b363dd29aa63dd2bd80cb45a001e37fe6/superdec/models/decoder.py#L32), simply wrapping decoder_layer in a deep copy should do the trick. Plus, this should keep your model definition compatible with the current checkpoints.

Cheers,
Noor
